### DISCUSSION TITLE
access-services-inside-options-and-startup-using-configureoptions/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#154] [Announcements]
# access-services-inside-options-and-startup-using-configureoptions/

This post shows how you can inject services when configuring IOptions implementations in your Startup class using the  IConfigureOptions interface.

https://andrewlock.net/access-services-inside-options-and-startup-using-configureoptions/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at January 08 2017, 10:02, in reply to [Dave Anderson](https://disqus.com/by/disqus_0naDR7TUri/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/154#discussioncomment-5795638)</em>

---

<p>The caching approach is still perfectly fine if it meets your requirements. This approach is better than number 1 from the Part 2-Caching article in the way it accesses an IServiceScopeFactory. If you want to go with number 2 (CachingAppTenantResolver), then you shouldn't need this approach.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 10 2017, 03:27, in reply to [Vladimir Georgiev](https://disqus.com/by/disqus_2ViHPhD56C/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/154#discussioncomment-5795642)</em>

---

<p>Thanks, really appreciate it!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at November 10 2017, 11:47, in reply to [Jonathan Channon](https://disqus.com/by/jchannon/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/154#discussioncomment-5795645)</em>

---

<p>Hmm, no, no idea! If the constructor is called it should definitely call Configure... Something strange going on there..</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at November 10 2017, 12:42, in reply to [Jonathan Channon](https://disqus.com/by/jchannon/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/154#discussioncomment-5795647)</em>

---

<p>The framework calls Configure when the service requests an instance of IOptions&lt;signalroptions&gt; in its constructor</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at November 10 2017, 01:55, in reply to [Jonathan Channon](https://disqus.com/by/jchannon/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/154#discussioncomment-5795651)</em>

---

<p>Try moving the Configure call after the AddSignalR call - order matters in ConfigureServices, so your values may be being overwritten by the defaults.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at June 15 2020, 04:56, in reply to [Sam](https://disqus.com/by/disqus_ZLQfirojSK/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/154#discussioncomment-5795671)</em>

---

<p>Hi, glad you enjoy them! My book doesn't really cover internals and edge cases like these - it is generally broader, and focuses more on the fundamentals as well.</p><p>To be honest the best way to learn thesec things is to try and build things, hit a problem, and then search for a problem. You'll generally find lots of posts like mine about it, or you can drive into the framework source code to find a solution (as I often do!)</p>

[a comment by an automated account is not shown]

**Comment by scuba6388** [comment]
Say I wanted to make an API call instead of using an EF context.  Am I OK to resolve an IHttpClientFactory in place of the IServiceScopeFactory in my IConfigureOptions implementation?  I would then like to grab an HttpClient from the factory in the Configure method in order to retrieve some settings.

**Reply by andrewlock** [reply]
> Am I OK to resolve an IHttpClientFactory in place of the IServiceScopeFactory in my IConfigureOptions implementation

To be safe, I would still resolve the `IServiceScopeFactory` in the ctor, create a scope, reolve the `HttpClient` or `IHttpClientFactory` from the service provider, make the request, and dispose the scope.

However, you're going to run into the fact that you can't use `async`/`await` here, and there's not a clean solution to that unfortunately
